### ISSUE TITLE
2 depth associations data using eager loading?

### ISSUE THREAD

**Issue by codeinus** [neo4jrb/neo4j#881]
I want to get all data related one model.

for example,
author has many posts and comments
post has many comments
each post and comment has one author
and I want to show all only 1 page
like this

| post.title | post.author | comments - comments.author |
| --- | --- | --- |
| abcd | a | 'good '- by b, 'so so' - by c |

can I get all data using a single query(with_associations())? 
Or although it is not a single query, how can I do that efficiently?
- also if relation between post and author has properties, can I get also properties of relation using with_associations()?

**Comment by cheerfulstoic** [comment]
Yes!  Make sure you're using the latest version of the `neo4j` gem first off.  Then you could do:

```ruby

Post.all.with_associations(:comments, :author).each do |post|
  puts post.title
  puts post.author.name
  post.comments.each do |comment|
    puts comment.body
  end
end
```

That would only execute a single query like this:

```cypher
OPTIONAL MATCH n<-[:`post`]-comments, n-[:`author`]->author WHERE (comments:Comment) AND (author:Person) RETURN n, [collect(comments),collect(author)]
```

If you don't include the `with_associations` it's supposed to eager load the `comments` and `author` associations with one query each (making three queries total), but looking at that it doesn't seem to be working...  I'll need to take a look at that, but let me know if `with_associations` works for you

**Comment by codeinus** [comment]
I did it, I got the results that I want
but it sends not a single query but many queries.
I want to get data 
 post -> author
 post -> comment
 post -> comment -> author (comment also has author)

query result is ...
```
Post.all.with_associations(:comments, :author).each do |post|
  post.title
  post.comments.each do |comment|
    comment.content
    comment.author.name
  end
end
 CYPHER 11ms MATCH (n:`Post`) OPTIONAL MATCH n<-[:`post`]-comments, n-[:`enrolled_in`]->author WHERE (comments:Comment) AND (author:Person) RETURN n, [collect(comments),collect(author)] 
 Comment#author 3ms MATCH comment1913, comment1913-[rel1:`author`]->(result_author:`Person`) WHERE (ID(comment1913) = {ID_comment1913}) RETURN result_author | {:ID_comment1913=>1913}
 Comment#author 3ms MATCH comment1912, comment1912-[rel1:`author`]->(result_author:`Person`) WHERE (ID(comment1912) = {ID_comment1912}) RETURN result_author | {:ID_comment1912=>1912}
 Comment#author 3ms MATCH comment1911, comment1911-[rel1:`author`]->(result_author:`Person`) WHERE (ID(comment1911) = {ID_comment1911}) RETURN result_author | {:ID_comment1911=>1911}
 Comment#author 3ms MATCH comment1908, comment1908-[rel1:`author`]->(result_author:`Person`) WHERE (ID(comment1908) = {ID_comment1908}) RETURN result_author | {:ID_comment1908=>1908}
 Comment#author 2ms MATCH comment1910, comment1910-[rel1:`author`]->(result_author:`Person`) WHERE (ID(comment1910) = {ID_comment1910}) RETURN result_author | {:ID_comment1910=>1910}
 Comment#author 3ms MATCH comment1909, comment1909-[rel1:`author`]->(result_author:`Person`) WHERE (ID(comment1909) = {ID_comment1909}) RETURN result_author | {:ID_comment1909=>1909}
[[], [#<Comment uuid: "80789d7d-e2e6-4eaa-a597-8ef25d4414f3", content: "not bad">, #<Comment uuid: "4da9789c-5e3b-4679-b860-2d3fb4122b1e", content: "not good">], [#<Comment uuid: "e3b4c3ad-96ed-4525-ad0c-cbc63c901f73", content: "no">], [], [#<Comment uuid: "218c8616-2c1b-4d0b-9fcc-b1867c92f66a", content: "good">, #<Comment uuid: "025da3c3-aafc-4d35-b3c6-286b39e1af2e", content: "so so">, #<Comment uuid: "c2475c50-1923-4a32-9d35-0c7b512b6ba3", content: "bad">], []]
```

also I want to see author in post
```
Post.all.with_associations(:comments, :author).each do |post|
  post.title
  post[:notes] # it is (author -[rel]- post)relation property
  post.author.name
  post.comments.each do |comment|
    comment.content
    comment.author.name
  end
end
 CYPHER 14ms MATCH (n:`Post`) OPTIONAL MATCH n<-[:`post`]-comments, n-[:`enrolled_in`]->author WHERE (comments:Comment) AND (author:Person) RETURN n, [collect(comments),collect(author)] 
NoMethodError: undefined method `name' for nil:NilClass
``` 

I`m using neo4j 5.0.5, neo4j-core 5.0.6
and I tested in rails console(Rails 4.1.8)

**Comment by cheerfulstoic** [comment]
Ah, I see the issue.  Currently eager loading with `with_associations` only supports one level deep of associations.

I think that the default eager loading should take care of getting all of the authors for the comments with a second query, though.  I'll have to have a look.  Thanks for pointing it out!

**Comment by cheerfulstoic** [comment]
Oh, duh, I see now that this was the title of the issue...

**Comment by codeinus** [comment]
+
if relation between post and author has properties, can I get also properties of relation when I`m using eager_loading?

**Comment by cheerfulstoic** [comment]
Nope, unfortunately not right now.  It might be good to have some mechanism for this and for associations in general, though I'm not sure of the best way about it.  The best way that I've thought of so far is a special method on the object that you get from the association, something like:

```ruby
Post.comments[0].followed_rel
```

**Comment by cheerfulstoic** [comment]
The other trick with that is that you don't necessarily always way to be loading that data if it's not efficient for what you're trying to do, so we'd want to be able to turn that on/off I think

**Comment by andrewhavens** [comment]
We were recently bit by this too:

```ruby
user.blogs.count
#> 2

user.blogs.map do |blog|
  blog.posts.count
end
#> [0, 0]

user.blogs.first.posts = [post1, post2]
user.blogs.map do |blog|
  blog.posts.count
end
#> [2, 2] # WRONG

user.blogs.with_associations(:posts).map do |blog|
  blog.posts.count
end
#> [2, 0] # Correct
```

**Comment by cheerfulstoic** [comment]
Oy, sorry about that.  I had some oddness recently as well.  This will be my first priority when I get a bit of time, thanks!

**Comment by cheerfulstoic** [comment]
FYI I've taken a look this evening and I understand the problem now.  It's going to take a bit to solve it, but I don't want to let this issue linger

**Comment by cheerfulstoic** [comment]
@andrewhavens I think I've made a fix which should help.  Could you try `5.0.13`?

**Comment by cheerfulstoic** [comment]
Any luck @andrewhavens et al?

**Comment by andrewhavens** [comment]
@cheerfulstoic Looks like we're using 5.2.5, so we probably upgraded a while ago? I don't think we're writing code like this anymore, so I don't necessarily have a way to test it at the moment. If you think you've fixed it, you probably did. I'd say you can close this issue.

**Comment by tjad** [comment]
Right now I'm not too concerned about the underlying query performance, but I would have the nested eager loading happen with the `with_associations` method, but allow the `with_associations` method to take a map object instead

From the example above,
Post.all.with_associations(:comments=>[:author])
should return all posts with their comments as well as the comments' author

So the handling should handle arrays and hashes/maps.

I am using neo4j gem 7.2.3 and neo4j 3.1.0
still can't eager load deep nested associations

**Comment by cheerfulstoic** [comment]
Thanks!  If you'd like to talk about the implementation or have any questions you might drop by our Gitter chat room: https://gitter.im/neo4jrb/neo4j

**Comment by thefliik** [comment]
@cheerfulstoic also closed by #1398, I think

**Comment by cheerfulstoic** [comment]
Yep, absolutely, thanks!  Anybody that's following this: This functionality was released in 8.2.0
